### ISSUE TITLE
LocalHook.CreateUnmanaged works while LocalHook.Create doesn't

### ISSUE THREAD

**Issue by etomm** [EasyHook/EasyHook#132]
The situation is simple. I create a process of QlikView suspended, and in it I inject easyhook. The injection hooks CreateFileW and CreateFileA using the LocalHook.Create. After it I resume QlikView and begin to command it trough COM calls. After awhile the process crashes with an access violation exception that I can't trap because it is inside ntdll.dll or clr.dll.

I commented the two LocalHook.Create calls and run again my tests and everything works. As soon as I reinsert the two hooks (also making them just call their hooked counterpart) the program return to crash in some minutes.

After this I tested the CreateUnmanaged in this way

LocalHook.CreateUnmanaged(
                    LocalHook.GetProcAddress("kernel32.dll", "CreateFileW"), 
                    Marshal.GetFunctionPointerForDelegate(_delegate), 
                    IntPtr.Zero);

That is slighty similar to what internally the LocalHook.Create does.

This made the process stay alive more than one hour.

I think the way the GCHandle.Alloc is used is wrong, because a Weak doesn't do anything good particularly: it allows the GC to collect the object and it doesn't pin it. So the pointer that you are reading inside the LocalHook.Create can soon point to memory no more owned from .net or can point to released memory.

I suppose your idea was to avoid that a LocalHook object would be collected if the other c# code would not keep a reference, but this is not the case if you use the Weak type and the Pinned type can be used just with particular objects.

**Comment by etomm** [comment]
I reviewed all my code and at the end I noticed in the test that the ThreadACL.SetExclusive was not called in the CreteUnmanaged one... so in fact I was never activating the hook.

Both fails the same way with an Access Violation.

To solve my problem I tried to hook the GetProcAddress and this makes everything fail really quickly the error code is this and is always this even cross restarts:

Faulting application name: Qv.exe, version: 11.20.13405.0, time stamp: 0x57728f17
Faulting module name: KERNELBASE.dll, version: 6.3.9600.18264, time stamp: 0x56e1bd71
Exception code: 0xe0434352
Fault offset: 0x0000000000008a5c
Faulting process id: 0x2798
Faulting application start time: 0x01d1f3b51626bb59
Faulting application path: C:\Program Files\QlikView\Qv.exe
Faulting module path: C:\Windows\system32\KERNELBASE.dll
Report Id: 54005615-5fa8-11e6-82fb-e82aeaaac3b8
Faulting package full name: 
Faulting package-relative application ID: 

The application is 64bit, the library that I load in the process is build in Debug in AnyCpu.

**Comment by etomm** [comment]
I tryed to load MinHook.x64 inside the process to use another library to put in place a hook and this one works better for GetProcAddress but has the exactly same issue for CreateFileW.

When hooking CreateFileW the process after some works crash down and the error in the event log is:
Faulting application name: Qv.exe, version: 11.20.13405.0, time stamp: 0x57728f17
Faulting module name: ntdll.dll, version: 6.3.9600.18233, time stamp: 0x56bb4ebb
Exception code: 0xc0000005
Fault offset: 0x00000000000546fb
Faulting process id: 0x20c4
Faulting application start time: 0x01d1f3d144a3a383
Faulting application path: C:\Program Files\QlikView\Qv.exe
Faulting module path: C:\Windows\SYSTEM32\ntdll.dll
Report Id: cb819421-5fc4-11e6-82fb-e82aeaaac3b8
Faulting package full name: 
Faulting package-relative application ID:

**Comment by spazzarama** [comment]
Are you doing anything in your hook? Or just calling original and returning? I would suggest testing when doing nothing first as a hook being in place doesn't by itself explain the behaviour u are experiencing

**Comment by etomm** [comment]
Inside the hook at the end to test I removed everything apart the call to the original method.

This kept crashing after sometime. The funny part is that that I'm testing now is that if I limit the hook to work on just one thread (in my case the main one) it seems working. Today I was going to perform more tests on this situation but yesterday the feeling was that or Easyhook or .NET itself doesn't support correctly calling a delegate from a unmanaged multithread environment. I even tried with a lock inside the hook to make just one call to the original method didn't work too.

**Comment by etomm** [comment]
Ok, I tested again the solution of limiting one thread (the main one seems the one interested from the CreateFileW that I must intercept). This allowed me to perform the same long running work for 5 times with success instead of 1 (it is a long procedure that copies object between one document and the other through COM, while the destination COM process is hooked with EasyHook).
To limit the thread I used the ThreadACL.SetInclusiveACL(<current thread id>) from inside the hook (because I don't know the thread id before). At the end the access violation.

Then If the file is the one that I need to change I change the path on the fly for EACH CreateFile that gots called with that file.

This lambda is the one that manage everything. It is based on a facade built around EasyHook and you can take it as a PSEUDOCODE.
`(string filename,
                        FileAccess access,
                        FileShare share,
                        IntPtr securityAttributes,
                        FileMode creationDisposition,
                        FileAttributes flagsAndAttributes,
                        IntPtr templateFile) =>
                    {

                        if (!filename.StartsWith(Constants.AppDataQlikViewBookmarksPath, StringComparison.OrdinalIgnoreCase) ||
                            !filename.EndsWith(".xml"))
                            return CreateFileW(filename, access, share, securityAttributes, creationDisposition, flagsAndAttributes, templateFile);

                        //this is the only thread that we care of
                        info.SetInclusiveAcl(info.GetCurrentThreadId());

                        var newFilename = Path.Combine(Constants.NPrintingQlikViewBookmarksPath, Path.GetFileName(filename.Insert(filename.Length - 4, _extensionPid)));
                        return CreateFileW(newFilename, access, share, securityAttributes, creationDisposition, flagsAndAttributes, templateFile);
}`

So it is no more the simple calling the original method from the hook. I take this as an evidence that the problem is not in my hook but somewhere around EasyHook or the .NET Framework. Seems like a concurrency problem that leaves a variable dirty.

At the end another access violation:
Faulting application name: Qv.exe, version: 11.20.13405.0, time stamp: 0x57728f17
Faulting module name: Qv.exe, version: 11.20.13405.0, time stamp: 0x57728f17
Exception code: 0xc0000005
Fault offset: 0x0000000000766a1d
Faulting process id: 0x2944
Faulting application start time: 0x01d1f47368072948
Faulting application path: C:\Program Files\QlikView\Qv.exe
Faulting module path: C:\Program Files\QlikView\Qv.exe
Report Id: 33339e60-6068-11e6-82fb-e82aeaaac3b8
Faulting package full name: 
Faulting package-relative application ID: 

If needed I can try to write a program that reproduce the problem, but not right now.

**Comment by etomm** [comment]
Is it a requirement for the hooked delegate to be STATIC?

**Comment by spazzarama** [comment]
No, I often don't use a static delegate and haven't had any issues.

**Comment by spazzarama** [comment]
Btw a reproduceable project for this issue would be much appreciated

**Comment by etomm** [comment]
I am able to reproduce it just with QlikView and a server connection. I will try to reproduce the problem with a local connection. Then, if you will install QlikView you should be able to verify it too. But right now I really have few time, I will try to do it in the next weeks

**Comment by etomm** [comment]
In the meantime that I will find a spare hour to write the complex test case, did you evaluate this? 

http://stackoverflow.com/questions/4097026/gchandle-to-get-addresspointer-of-net-object

As I said at the opening, you use GCHandle.Alloc to pass the LocalHook to the native hook, but the address that is returned is not guaranted to remain the same.

**Comment by hemantabhagat** [comment]
Hi, did you solved the issue? I am facing the same issue when I inject custom CreateProcess to make clean process.
